### ISSUE TITLE
Async ServiceDiscovery not refreshing records

### ISSUE THREAD

**Issue by brennie** [balliegojr/simple-dns#15]
I have a setup with two async ServiceDiscovery on the same host with a TTL of 30s. Consistently, `get_known_services()` will eventually start returning an empty list for each service. I added some debug logging and it appears to be the case that the main executor loop can sleep *past* the refresh time of a resource. Then, when we go to [refresh our known instances](https://github.com/balliegojr/simple-dns/blob/15d9e335b4cbe529c1f652deb60f073593ce79d4/simple-mdns/src/async_discovery/service_discovery.rs#L234):

```rust
    async fn refresh_known_instances(&self) -> std::io::Result<Instant> {
        log::info!("Refreshing known services");
        let now = Instant::now();
        let next_expiration = self
            .resource_manager
            .read()
            .await
            .get_next_expiration() // <--
            .map(Instant::from_std);
```

we call [get_next_expiration](https://github.com/balliegojr/simple-dns/blob/15d9e335b4cbe529c1f652deb60f073593ce79d4/simple-mdns/src/resource_record_manager.rs#L73):

```rust
pub fn get_next_expiration(&self) -> Option<Instant> {
        self.resources
            .iter()
            .flat_map(|(_, resources)| {
                resources.values().filter_map(|resource_type| {
                    if resource_type.should_refresh() { // <--
                        return None;
                    }
                    match resource_type {
                        ResourceRecordType::Owned => None,
                        ResourceRecordType::Expirable(exp_info) => Some(exp_info.refresh_at),
                    }
                })
            })
            .min_by(|a, b| a.cmp(b))
    }
```

However, if we have slept *past* a resource's refresh time, `should_refresh()` will return False:

```rust
    pub fn should_refresh(&self) -> bool {
        match self {
            ResourceRecordType::Owned => false,
            ResourceRecordType::Expirable(exp_info) => exp_info.refresh_at < Instant::now(), // <--
        }
    }
```

Commenting out the call to `should_refresh()` fixes my issue.

Here is a minimal worked example of the issue:

```rust
use std::net::SocketAddr;
use std::str::FromStr;
use std::time::{Duration, Instant};

use simple_mdns::async_discovery::ServiceDiscovery;
use tokio::time::interval;

#[tokio::main]
async fn main() {
    let mut sda = ServiceDiscovery::new("sda", "_example._tcp.local", 30).unwrap();
    let mut sdb = ServiceDiscovery::new("sdb", "_example._tcp.local", 30).unwrap();

    sda.add_service_info(SocketAddr::from_str("128.0.0.1:1234").unwrap().into()).await.unwrap();
    sdb.add_service_info(SocketAddr::from_str("127.0.0.1:4321").unwrap().into()).await.unwrap();

    let mut timer = interval(Duration::from_secs(3));
    let start = Instant::now();

    loop {
        timer.tick().await;

        println!("{}:", start.elapsed().as_secs());
        println!("  A: {:?}", sda.get_known_services().await);
        println!("  B: {:?}", sdb.get_known_services().await);
        println!("---")
    }
}
```

Output:
```
0:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
3:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
6:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
9:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
12:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
15:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
18:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
21:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
24:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
27:
  A: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [4321], attributes: {"": None} }]
  B: [InstanceInformation { ip_addresses: [127.0.0.1], ports: [1234], attributes: {"": None} }]
---
30:
  A: []
  B: []
---
33:
  A: []
  B: []
---
```

**Comment by balliegojr** [comment]
Thank you very much for reporting this.

The call to `should_refresh` is actually inverted, the function have to return `None` when it shouldn't refresh
```rust
pub fn get_next_expiration(&self) -> Option<Instant> {
        self.resources
            .iter()
            .flat_map(|(_, resources)| {
                resources.values().filter_map(|resource_type| {
                    if !resource_type.should_refresh() { // <-- this is missing the ! 
                        return None;
                    }
                    match resource_type {
                        ResourceRecordType::Owned => None,
                        ResourceRecordType::Expirable(exp_info) => Some(exp_info.refresh_at),
                    }
                })
            })
            .min_by(|a, b| a.cmp(b))
    }
```

Your issue reporting skills are amazing BTW

**Comment by balliegojr** [comment]
I have published a new version, please let me know if it fixed the issue

**Comment by brennie** [comment]
Thank you! And thank you for the quick turnaround!

I can confirm this issue is fixed in 0.4.1.

**Comment by balliegojr** [comment]
Nice, thanks!
I'll close this issue then.
